File: Java/More on Classes.java
// 41 Lesson

//Drag and drop from the options below to create a valid Java code with encapsulation.
public class Person {				
 
 private int age;
 
 public void setAge(int age) {

     if (age > 0) {

        this.age = age;

      }

   }

}

// 42 Lesson

//Fill in the blank to inherit the Car class from the Vehicle class.
class Car extends Vehicle { }

//Fill in the blanks to inherit from the Animal class and call its method in main.
class Animal {

   public void makeSound() {

   System.out.println(''Hi'');

   }

}

class Dog extends
 Animal {

}

class A {

   public static void main(String args[ ]) {

   Dog dog = new Dog();
   
   dog.makeSound();

   }

}

//Private methods are inherited from the super class.

//False

// 43 Lesson

//Briefly, polymorphism is...

//... one method, with different implementations

// 44 Lesson

//Overridden methods should have the same return type and arguments as the parent method.

//True

/*
What is the output of this code?
class A {
   public void doSomething() {
     System.out.println("A");
   }
   public void doSomething(String str) {
     System.out.println(str);
   }
}
class B {
   public static void main(String[ ] args) {
   A object = new A();
   object.doSomething("B");
   }
}
*/

//B

// 45 Lesson

//A class containing an abstract method is an abstract class.

//True

//Fill in the blanks to create an abstract class with an abstract method and inherit from it.

abstract class Animal {

  public int age;

  public abstract int printAge();

}

class Dog extends Animal {

  public int printAge() {

    return age;

  }	

}

// 46 Lesson

//In Java, how many superclasses can your inherited subclass have?

//only one

//Drag and drop from the options below to implement an interface.
interface Animal {

   public void eat();

}

class Cat implements Animal {

   public void eat() {

      System.out.println("Cat eats");

   }

}


// 47 Lesson

/*
What is the output of this code?
public static void main(String[ ] args) {	
  double x = 1.5;
  double y = 2.65;
  sum((int)x, (int)y);
}
static void sum(int x, int y) {
   System.out.println(x + y);
}
*/

//3


// 48 Lesson

/*
What is the output of this code?
class A {
   public void print() {
      System.out.println("A");
   }
}
class B extends A {
   public void print() {
      System.out.println("B");
   }
   public static void main(String[ ] args) {
    A object = new B();
    B b = (B) object;
    b.print();
   }
}
*/

//B

// 49 Lesson

//Fill in the blanks to override the start method of the Machine class.
Machine m = 
new Machine() {

   
@Override public void start() {

      System.out.println("Hi");

   }

}

//Drag and drop from the options below to print "Hello".
class A {

   public void print() {

   System.out.println("A");

   }		

}

class B {				

   public static void main(String[ ] args) {	
 
   A object = new A() {


      @Override public void print() {

       System.out.println("Hello");

       }

    };

    object.print();

   }		

}

// 50 Lesson

//Rearrange the code to have an inner class Hand, which has a method called "shake" that prints "Hi".

public class Person {
class Hand{
	public void shake(){
	System.out.println("Hi");}
}}

// 51 Lesson

/*
What is the output of this code?
class A {
 private int x;	
 public static void main(String[ ] args) {
   A a = new A();
   a.x = 5;
   A b = new A();
   b.x = 5;
   System.out.println(a == b);			
 }
}
*/

//false

//Drag and drop from the options below to check whether the two objects of type A are semantically equal.
class A {

   private int x;

   public boolean equals(Object o) {
    
   return ((A)o).x == this.x;
   }	

   public static void main(String[ ] args)  {

    A a = new A();

    a.x = 9;

    A b = new A();

    b.x = 5;

    System.out.println(a.equals(b));

   }

}


// 52 Lesson

//Enums are used to declare variables that represent...

//members of a fixed set

//Drag and drop from the options below to create an Enum called "Color", with the values RED, BLUE, GREEN.
public enum Color  {

  RED, BLUE, GREEN;

}

/*
What is the output of this code?
enum Color  {
  RED, BLUE, GREEN;
}
class PrintColor {
   public static void main(String[ ] args) {
   Color color = Color.RED; 
   switch(color) {
   case BLUE:
     System.out.println("1");
   break;
   case GREEN:
     System.out.println("2");
   break;
   default:
     System.out.println("0");
   break;		 
   }
}
}
*/

//0
 
 
// 53 Lesson

//Fill in the blank to import all types in the package awt.
import java.awt.*;


// 54 Lesson

/*
What is the output of this code?
class A {
   private void print() {
     System.out.println(''a'');
   }
   private void print(String str) {
     System.out.println(''b'');
   }
   private void print(int x) {
     System.out.println(''c'');
   }		
   public static void main(String[ ] args) {	
     A object = new A();
     object.print(12);	
   }	
}
*/

//c

//Fill in the blanks to define a new class Falcon, based on the superclass Bird.
class Falcon extends Bird {

 }

//Object variables store...

//References

//What term is used for hiding the details of an object from the other parts of a program?

//Encapsulation

//A class Car and its subclass BMW each have a method run(), which was written by the developer as part of the class definition. If CarObj refers to an object of type BMW, what will CarObj.run(); do?

//The run() method defined in BMW will be called.

/*
Valentine, Holiday, and Birthday inherit from the class Card. In order for the following code to be correct, what type must the reference variable card be?
card = new Valentine( "A", 14 ) ;
card.greeting();

card = new Holiday( "B" ) ;
card.greeting();

card = new Birthday( "C", 12 ) ;
card.greeting();
*/

//Card

// 55 Lesson

/*
Shapes


You are working on a graphical app, which includes multiple different shapes.
The given code declares a base Shape class with an abstract area() method and a width attribute.
You need to create two Shape subclasses, Square and Circle, which initialize the width attribute using their constructor, and define their area() methods.
The area() for the Square class should output the area of the square (the square of the width), while for the Circle, it should output the area of the given circle (PI*width*width).
The code in main creates two objects with the given user input and calls the area() methods.

Sample Input:
5
2

Sample Output:
25
12.566370614359172

The area of the square is 5*5=25, while the area of the circle is PI*2*2=12.566370614359172
*/
import java.util.Scanner;

abstract class Shape {
    int width;
    abstract void area();
}

class Square extends Shape{
    public Square(int x){
        width=x;
    }
    void area(){
    System.out.println(width*width);
    }
}

class Circle extends Shape{
    public Circle(int x){
        width=x;
    }
    void area(){
    System.out.println(Math.PI*width*width);
    }
}

public class Program {
    public static void main(String[ ] args) {
        Scanner sc = new Scanner(System.in);
        int x = sc.nextInt();
        int y = sc.nextInt();
        
        Square a = new Square(x);
        Circle b = new Circle(y);
        a.area();
        b.area();
    }
}

File: C#/More On Classes.cs
// 50 Lesson

//Drag and drop from the options below to declare a destructor for class Person.

class Person {

   ~Person(){

  }

}

//Destructors are:

//invoked when an object is deleted

// 51 Lesson

//Drag and drop from the options below to have a valid Main method:
 

static void Main(string[] args)

//Drag and drop from the options below to assign the value of the Math class static PI variable to x.

double x = Math.PI;

//To be able to directly call a method in Main, it should be:

//static

//Drag and drop from the options below to initialize the static Age property of the Person class using a static constructor:

class Person {

  public static int Age { get; set; }

  static Person() {

    Age = 0;

  }

}

// 52 Lesson

//A static class contains:

//only static members

/*
What is the output of this code?
int[] arr = { 9, 4, 10, 42 };
Array.Sort(arr);
double x = Math.Sqrt(arr[1]);
Console.WriteLine(x);
*/

//3

// 53 Lesson

//Fill in the blanks to print the age variable of the Person class using the this keyword.

class Person {

  private int age;

  public void Print() {

    Console.Write(this.age);

  }

}

//Which statements are true?
//Select All That Apply

//constants should be assigned a value when declared
//readonly fields can be initialized by the constructor

// 54 Lesson

/*
What is the output of this code?
string str = "I know C#";
char x = str[3];
Console.WriteLine(x);
*/

//n

//Fill in the blanks to define a valid indexer:

class Person {

  private string name;

  public char this[int index] {

    get {

      return name[index];

    }

  }

}


// 55 Lesson

//Operator overloading means:

//defining custom actions for operators

//Fill in the blanks to overload the greater than operator for the Box class.

public static bool operator>(Box a,Box b) {

  if (a.Height*a.Width > b.Height*b.Width)

    return true;

  else

    return false;

}


// 56 Lesson

//Fill in the blanks to declare a destructor for the class Book.

class Book {

  ~Book() {

    //some code

  }

}

/*
What is the output of this code?
class Temp {
  public int num = 2;
  public Temp() { num++;}
  ~Temp() { num++; }
}
static void Main(string[] args) {
  Temp t = new Temp();
  Console.WriteLine(t.num);
}
*/

//3

//Fill in the blanks to make a valid working program.

static void Func()
 {

  Console.Write("Hi there");

} 

static void Main(string[] args) {

  Func();

}

/*
What is the value of x after this code?
int[] arr = {8, 5, 4};
Array.Reverse(arr);
double x = Math.Pow(arr[0], 2);
*/

//16

//Fill in the blanks to define a valid indexer:

public int this[int index] {

  get {

    return n[index];

  }

  set {

    n[index] = value;

  }

}

//Fill in the blanks to overload the minus operator for the T class:

public static T operator- (T a, T b) {

  T res = new T(a.num - b.num);

  return res;

}

// 57 Lesson

/*
Dance


In a ballroom dancing competition, each dancer from a pair is evaluated separately, and then their points are summed up to get the total pair score.
The program you are given takes the names and the points of each dancer as input and creates a DancerPoints objects for each dancer, using the taken name and score values as parameters for constructors.
Complete the given class, using overload + operator to return an new object where the names of dancers are in one string (see sample output) and the score is equal to the sum of their points.
The declaration of that object and the output of its points are already written in Main().

Sample Input
Dave
8
Jessica
7

Sample Output
Dave & Jessica
15
*/

using System;
using System.Collections.Generic;

namespace Code_Coach_Challenge
{
    class Program
    {
        static void Main(string[] args)
        {
            string name1 = Console.ReadLine();
            int points1 = Convert.ToInt32(Console.ReadLine());
            string name2 = Console.ReadLine();
            int points2 = Convert.ToInt32(Console.ReadLine());

            DancerPoints dancer1 = new DancerPoints(name1, points1);
            DancerPoints dancer2 = new DancerPoints(name2, points2);

            DancerPoints total = dancer1 + dancer2;
            Console.WriteLine(total.name);
            Console.WriteLine(total.points);
        }
    }

    class DancerPoints
    {
        public string name;
        public int points;
        public DancerPoints(string name, int points)
        {
            this.name = name;
            this.points = points;
        }

        public static DancerPoints operator +(DancerPoints dancer1,DancerPoints dancer2)
            => new DancerPoints(dancer1.name + " & " + dancer2.name, dancer1.points+dancer2.points);

    }
}